### ISSUE TITLE
[Question] 48G Finetuning Out Of Memory

### ISSUE THREAD

**Issue by zhl98** [haotian-liu/LLaVA#266]
### Question

48G graphics memory fine-tuning 7B model Out Of Memory.
My cmd is: I use the checkpoint  liuhaotian/LLaVA-Lightning-MPT-7B-preview.

`CUDA_VISIBLE_DEVICES=2 torchrun --nnodes=1 --nproc_per_node=1 --master_port=25009  train_mem.py --model_name_or_path ./checkpoints/llava-7b-mpt/ --version v1  --data_path ./playground/scienceqa/train_true.json --image_folder ./playground/scienceqa/image/  --vision_tower ./checkpoints/clip-large --mm_vision_select_layer -2  --bf16 True --output_dir ./checkpoints/llava-7b-pretrain-no_im_start_end_token-finetune_scienceqa --num_train_epochs 12 --per_device_train_batch_size 1  --per_device_eval_batch_size 1  --gradient_accumulation_steps 1  --evaluation_strategy "no" --save_strategy "steps"  --save_steps 5000 --save_total_limit 3 --learning_rate 2e-5  --weight_decay 0.  --warmup_ratio 0.03  --lr_scheduler_type "cosine"   --logging_steps 1 --tf32 True --fsdp "full_shard auto_wrap"  --fsdp_transformer_layer_cls_to_wrap 'MPTBlock'  --model_max_length 2048  --gradient_checkpointing True  --lazy_preprocess True  --report_to wandb`

**Comment by BigJoon** [comment]
Did you solve it? Have you checked the cpu memory as well?
If cpu memory is the problem, try increasing the swap memory size.

**Comment by zhl98** [comment]
I haven't solved this problem. Three 48g graphics cards are required to fine-tune this model. I am unable to use the deepspeed module for acceleration, resulting in a large and slow existing system.

**Comment by haotian-liu** [comment]
@zhl98 

Hi, I am a bit confused about the request, are you trying to finetune the model with only a single GPU or with multiple GPUs but failed? We've tested that 4x A6000s work for finetuning. Thanks.

**Comment by zhl98** [comment]
Yes, I can make fine adjustments using three A6000 graphics cards, but not two. I tried to use deepspeed to reduce the occupation of graphics memory, but failed.

**Comment by haotian-liu** [comment]
@zhl98 

Two possibilities if you really want to squeeze your process into two GPUs.

1. Use [cpu offloading](https://github.com/haotian-liu/LLaVA/blob/main/scripts/zero3_offload.json).
2. Use [QLora](https://github.com/haotian-liu/LLaVA/blob/main/scripts/finetune_qlora.sh)

Both shall let you train on two A6000s, and you can trade-off between speed/accuracy on your own machine :)

**Comment by zhl98** [comment]
Thank you very much for your answer.

**Comment by HashmatShadab** [comment]
Will cpu offloading also lead to performance degradation or just decrease the speed?
